auto_round/utils/device.py

```diff
@@ -732,6 +731,75 @@ def find_best_device(layer_name, estimated_memory, layer_idx):
     return ordered_device_map, names
 
 
+def get_moe_memory_ratio(block: torch.nn.Module) -> float:
+    """
+    Calculate the memory ratio for MoE (Mixture of Experts) models.
+
+    For MoE models, only num_experts_per_tok experts are activated per token,
+    not all experts. This function returns the ratio of active experts to total experts.
+
+    Args:
+        block (torch.nn.Module): The model block to analyze.
+
+    Returns:
+        float: Memory ratio (num_experts_per_tok / num_experts).
+               Returns 1.0 for non-MoE models.
+        bool: True if the model is MoE, False otherwise.
+
+    Examples:
+        - Non-MoE model: returns 1.0
+        - Mixtral (2/8 experts): returns 0.25
+        - Qwen2MoE (4/60 experts): returns ~0.067
+    """
+    from auto_round.utils.model import is_moe
+
+    for name, module in block.named_modules():
+        if is_moe(module):
+            config = getattr(block, "config", None)
+            if config is None:
+                break
+
+            # Try to get num_experts_per_tok (active experts count)
+            num_experts_per_tok = getattr(
+                config, "num_experts_per_tok", None
+            )  # Mixtral, Qwen2MoE, DeepSeek, GPT-OSS, Llama4, LLaDAMoE
+            if num_experts_per_tok is None:
+                num_experts_per_tok = getattr(config, "moe_num_active_primary_experts", None)  # SmallThinker
+            if num_experts_per_tok is None:
+                # HunYuan MoE uses moe_topk (array), get first element
+                moe_topk = getattr(config, "moe_topk", None)  # HunYuan MoE V1
+                if moe_topk is not None and isinstance(moe_topk, (list, tuple)) and len(moe_topk) > 0:
+                    num_experts_per_tok = moe_topk[0]
+                elif moe_topk is not None:
+                    num_experts_per_tok = moe_topk
+
+            if num_experts_per_tok is None:
+                break
+
+            # Get total number of experts
+            num_experts = getattr(config, "num_local_experts", None)  # Mixtral, PhiMoE, Grok, Llama4
+            if num_experts is None:
+                num_experts = getattr(
+                    config, "num_experts", None
+                )  # Qwen2MoE, Olmo, BailingMoE, GroveMoE, HunYuan, LLaDAMoE
+            if num_experts is None:
+                num_experts = getattr(config, "moe_num_primary_experts", None)  # SmallThinker
+            if num_experts is None:
+                num_experts = getattr(config, "n_routed_experts", None)  # DeepSeek
```

Comment:
if num_experts is none, better use all the linear number in the block

Reply:
Yes, if any required value is None, we set moe_ratio=1.0 meaning all linear is considered.